File: src/test/kotlin/samlang/cli/ConfigurationParserTest.kt
package samlang.cli

import io.kotlintest.shouldBe
import io.kotlintest.shouldThrow
import io.kotlintest.specs.StringSpec

class ConfigurationParserTest : StringSpec() {
    init {
        "empty configuration can parse." {
            parseConfiguration(string = "{}") shouldBe samlang.Configuration(
                sourceDirectory = ".",
                outputDirectory = "out",
                excludes = emptyList(),
                targets = emptySet()
            )
        }
        "partial configuration can parse." {
            val json = """{ "sourceDirectory": "source", "targets": ["ts", "js", "java"] }"""
            parseConfiguration(string = json) shouldBe samlang.Configuration(
                sourceDirectory = "source",
                outputDirectory = "out",
                excludes = emptyList(),
                targets = setOf("ts", "js", "java")
            )
        }
        "full configuration can parse." {
            val json = """
            {
                "sourceDirectory": "source",
                "outputDirectory": "output",
                "excludes": ["foo", "bar"],
                "targets": ["ts", "js", "java"]
            }
            """.trimIndent()
            parseConfiguration(string = json) shouldBe samlang.Configuration(
                sourceDirectory = "source",
                outputDirectory = "output",
                excludes = listOf("foo", "bar"),
                targets = setOf("ts", "js", "java")
            )
        }
        "empty string does not parse." {
            shouldThrow<IllFormattedConfigurationException> { parseConfiguration(string = "") }
        }
        "bad json does not parse." {
            shouldThrow<IllFormattedConfigurationException> { parseConfiguration(string = "{") }
            shouldThrow<IllFormattedConfigurationException> { parseConfiguration(string = "}") }
        }
        "bad format does not parse." {
            shouldThrow<IllFormattedConfigurationException> {
                parseConfiguration(string = """{ "sourceDirectory": 3 }""")
            }
            shouldThrow<IllFormattedConfigurationException> {
                parseConfiguration(string = """{ "outputDirectory": 3 }""")
            }
            shouldThrow<IllFormattedConfigurationException> {
                parseConfiguration(string = """{ "excludes": 3 }""")
            }
            shouldThrow<IllFormattedConfigurationException> {
                parseConfiguration(string = """{ "targets": 3 }""")
            }
            shouldThrow<IllFormattedConfigurationException> {
                parseConfiguration(string = """{ "excludes": [3] }""")
            }
            shouldThrow<IllFormattedConfigurationException> {
                parseConfiguration(string = """{ "targets": [3] }""")
            }
            shouldThrow<IllFormattedConfigurationException> {
                parseConfiguration(string = """{ "excludes": ["3", 4] }""")
            }
            shouldThrow<IllFormattedConfigurationException> {
                parseConfiguration(string = """{ "targets": ["3", 4] }""")
            }
        }
        "bad target does not parse." {
            shouldThrow<IllFormattedConfigurationException> {
                parseConfiguration(string = """{ "targets": ["3"] }""")
            }
        }
    }
}

File: typescript-rewrite/samlang-cli/src/configuration/__tests__/configuration-parser.test.ts
import parseSamlangProjectConfiguration from '../configuration-parser';

it('Parser can parse good configurations', () => {
  expect(parseSamlangProjectConfiguration('{}')).toEqual({
    sourceDirectory: '.',
    outputDirectory: 'out',
    excludes: [],
  });

  expect(parseSamlangProjectConfiguration('{"sourceDirectory": "source"}')).toEqual({
    sourceDirectory: 'source',
    outputDirectory: 'out',
    excludes: [],
  });

  expect(parseSamlangProjectConfiguration('{"outputDirectory": "out-out"}')).toEqual({
    sourceDirectory: '.',
    outputDirectory: 'out-out',
    excludes: [],
  });

  expect(
    parseSamlangProjectConfiguration(`{
      "sourceDirectory": "source",
      "outputDirectory": "output",
      "excludes": ["foo", "bar"]
    }
    `)
  ).toEqual({
    sourceDirectory: 'source',
    outputDirectory: 'output',
    excludes: ['foo', 'bar'],
  });
});

it('Parser can reject bad configurations', () => {
  expect(parseSamlangProjectConfiguration('')).toBeNull();
  expect(parseSamlangProjectConfiguration('null')).toBeNull();
  expect(parseSamlangProjectConfiguration('undefined')).toBeNull();
  expect(parseSamlangProjectConfiguration('1')).toBeNull();
  expect(parseSamlangProjectConfiguration('"undefined"')).toBeNull();
  expect(parseSamlangProjectConfiguration('{')).toBeNull();
  expect(parseSamlangProjectConfiguration('}')).toBeNull();
  expect(parseSamlangProjectConfiguration('{ "sourceDirectory": 3 }')).toBeNull();
  expect(parseSamlangProjectConfiguration('{ "outputDirectory": 3 }')).toBeNull();
  expect(parseSamlangProjectConfiguration('{ "excludes": 3 }')).toBeNull();
  expect(parseSamlangProjectConfiguration('{ "excludes": [3] }')).toBeNull();
  expect(parseSamlangProjectConfiguration('{ "excludes": [3, "4"] }')).toBeNull();
});
